minio_storage/storage.py

```diff
@@ -88,57 +93,54 @@ def _save(self, name, content):
                                    content_size,
                                    content_type)
             return sane_name
-        except ResponseError as error:
-            logger.warn(error)
-            raise IOError("File {} could not be saved".format(name))
+        except merr.ResponseError as error:
+            raise minio_error("File {} could not be saved".format(name), error)
 
     def delete(self, name):
         # type: (str) -> None
         try:
             self.client.remove_object(self.bucket_name, name)
-        except ResponseError as error:
-            logger.warn("Object deletion failed")
-            logger.warn(error)
-            raise IOError("Could not remove file {}".format(name))
+        except merr.ResponseError as error:
+            raise minio_error("Could not remove file {}".format(name), error)
 
     def exists(self, name):
         # type: (str) -> bool
         try:
             self.client.stat_object(
                 self.bucket_name, self._sanitize_path(name))
             return True
-        except ResponseError as error:
+        except merr.ResponseError as error:
             # TODO - deprecate
             if error.code == "NoSuchKey":
                 return False
             else:
-                logger.warn(error)
-                raise IOError("Could not stat file {}".format(name))
-        except NoSuchKey as error:
-            return False
-        # Temporary - due to https://github.com/minio/minio-py/issues/514
-        except NoSuchBucket as error:
+                raise minio_error("Could not stat file {}".format(name), error)
+        except merr.NoSuchKey as error:
             return False
+        except merr.NoSuchBucket:
+            raise
         except Exception as error:
-            logger.warn(error)
+            logger.error(error)
 
     def listdir(self, prefix):
         try:
             # TODO: break the path
             objects = self.client.list_objects(self.bucket_name, prefix)
             return objects
-        except ResponseError as error:
-            logger.warn(error)
-            raise IOError("Could not list directory {}".format(prefix))
+        except merr.NoSuchBucket:
+            raise
```

Comment:
I think this shall not raise to outside of Storage member call.
What we raise shall be all inherited out of MinIOError.
It also can inherit then from NoSuchBucket for those who really want to see it, but for 
everybody else all errors shall be easily catchable with 
except MinIOError
without need to learn some lower level package
(minio-py).

Reply:
I'd like configuration errors to sip up as e500 errors as early as possible instead of risking having a service running which just logs errors instead of clearly signaling that something is wrong.

Reply:
If anything I would maybe consider wrapping those errors in in djangos ImproperlyConfigured error instead if just re raising them but that also hides details from app/log monitoring systems so I'm not entirely convinced about the usefulness

Reply:
I just started off with the list of exceptions which will be reraised without
wrapping here and it's a good enough start:
https://github.com/py-pa/django-minio-storage/pull/40/files#diff-4da1f143e40334ab933cf89bbb1eab47R12

The line you mentoned above should probably have a raise minio_error() instead
but I didnt have much time yesterday to properly manually test a lot of
scenarios so I decided to avoid making modifications where behaviour could
possibly change without me being able to properly verify it.

Reply:
Probably ImproperlyConfigured makes sense.
At least, it is something that is common and we easily name it as part of our interface.
Not something completely unknown buried in third component user does not suspect even existed.

Reply:
I will document the exception behavior instead of wrapping everything and maybe wrap some errors as  ImproperlyConfigured where it makes sense.